### PR TITLE
feat: auto login session will expire after 24h

### PR DESCRIPTION
**Pull request by Kininaru**
Fix: https://github.com/casbin/casdoor/issues/109

Deployed at: https://door.chromium.link

Video: (Modified database to simulate login expiration)

https://user-images.githubusercontent.com/42638489/124791965-259fc380-df7f-11eb-84f7-8209a256ac24.mp4

### CODE REVIEW

**Comment by Kininaru** [comment]
@hsluoyz plz review

File: object/user.go [hunk 1]
```diff
@@ -22,10 +22,11 @@ import (
...
 type User struct {
-	Owner       string `xorm:"varchar(100) notnull pk" json:"owner"`
-	Name        string `xorm:"varchar(100) notnull pk" json:"name"`
-	CreatedTime string `xorm:"varchar(100)" json:"createdTime"`
-	UpdatedTime string `xorm:"varchar(100)" json:"updatedTime"`
+	Owner           string `xorm:"varchar(100) notnull pk" json:"owner"`
+	Name            string `xorm:"varchar(100) notnull pk" json:"name"`
+	CreatedTime     string `xorm:"varchar(100)" json:"createdTime"`
+	UpdatedTime     string `xorm:"varchar(100)" json:"updatedTime"`
+	LoginExpireTime string `xorm:"varchar(100)" json:"loginExpireTime"`
```
Comment (hsluoyz) [inline object/user.go:29]: Login -> Signin

File: controllers/account.go [hunk 2]
```diff
@@ -52,6 +52,7 @@ type RequestForm struct {
 	EmailCode   string `json:"emailCode"`
 	PhoneCode   string `json:"phoneCode"`
 	PhonePrefix string `json:"phonePrefix"`
+	Remember    bool   `json:"remember"`
```
Comment (hsluoyz) [inline controllers/account.go:55]: Remember -> AutoSignin

**Review by hsluoyz — CHANGES_REQUESTED** [review]

File: controllers/base.go [hunk 3]
```diff
@@ -26,6 +32,24 @@ func (c *ApiController) GetSessionUser() string {
...
+		return ""
+	}
+
+	userObj := object.GetUser(userId)
+	if userObj == nil {
+		return ""
+	}
+
+	// if user login expired, then clean the session and return an empty string
+	if userObj.LoginExpireTime != "" && userObj.LoginExpireTime < util.GetCurrentTime() {
```
[the remark this answers is not in the document]
Comment (Kininaru) [reply controllers/base.go:46]: `LoginExpireTime` and `Util.GetCurrentTime` are strings like `2021-06-07T21:45:58+08:00`. 

`year - month - day - hour - minute - second`

We can use `<` to compare the ASCII of the first different char between 2 time strings to tell which is earlier.

File: object/user.go (hunk 1 above)
Comment (Kininaru) [reply object/user.go]: @hsluoyz Fixed.

File: controllers/account.go (hunk 2 above)
Comment (Kininaru) [reply controllers/account.go]: @hsluoyz Fixed.

File: controllers/base.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (hsluoyz) [reply controllers/base.go]: @Kininaru but what about the timezone?

+08:00 > +07:00 (may be right)
-08:00 > -07:00 (wrong!)

You should use a formal date comparison in Go.

[the remark this answers is not in the document]
Comment (hsluoyz) [reply controllers/base.go]: @Kininaru

[the remark this answers is not in the document]
Comment (Kininaru) [reply controllers/base.go]: Got that, I am going to use Unix time stamp instead of this, is that OK?

[the remark this answers is not in the document]
Comment (Kininaru) [reply controllers/base.go]: @hsluoyz Fixed that, wrote a new function:
```go
func IsTimeExpired(timeString string) bool {
	t, err := time.Parse(time.RFC3339, timeString)
	if err != nil {
		panic(err)
	}

	return t.Unix() < time.Now().Unix()
}
```

File: controllers/base.go [hunk 4]
```diff
@@ -26,6 +32,24 @@ func (c *ApiController) GetSessionUser() string {
...
+		return ""
+	}
+
+	userObj := object.GetUser(userId)
+	if userObj == nil {
+		return ""
+	}
+
+	// if user login expired, then clean the session and return an empty string
+	if userObj.SigninExpireTime != "" && util.IsTimeExpired(userObj.SigninExpireTime) {
```
Comment (hsluoyz) [inline controllers/base.go:46]: What if the user logoff and then login?

What if he has multiple sessions?

**Review by hsluoyz — CHANGES_REQUESTED** [review]

File: controllers/base.go (hunk 4 above)
Comment (Kininaru) [reply controllers/base.go:46]: These codes change the session, and is not that strongly related to accounts.

If a user has many sessions, each of the session is independent. Sessions with "auto sign in" won't expire, and sessions without "auto sign in" will expire.

Comment (hsluoyz) [reply controllers/base.go:46]: But won't `object.UpdateUserInternal(userId, userObj)` write to DB?

Comment (Kininaru) [reply controllers/base.go:46]: @hsluoyz Oh, that's right... I did not take this into consideration.

Can I add a time stamp to the session, so the session will become something like this: `built-in/admin 1626358811`

Then add some code to `GetSessionUser` to recognize the time stamp.

Comment (Kininaru) [reply controllers/base.go]: @hsluoyz Resolved that. Now, exipre timestamps are stored in session.

Checked auto signin:


Did not check:
$( _LRYFMW4ACOUVB4OES5N

Comment (hsluoyz) [reply controllers/base.go]: @Kininaru  We can store a struct in the session, you can see Casnode's code. We don't need to put all stuff in one string and then parse it. This code is fragile..

Comment (Kininaru) [reply controllers/base.go]: @hsluoyz Can I add another session called `ExpireTime`? Or store a struct in a session called `SessionInfo`...
I found that many places use the session `username` like `built-in/admin`, turning the session to a struct (or even just add a timestamp behind user id) may cause many unexpected bugs...

Comment (hsluoyz) [reply controllers/base.go]: @Kininaru we can extract `GetSessionUsername()` first before this PR like this:

https://github.com/casbin/casnode/blob/9c66ea1de047a22c9d99e36bcd86dbc9280acb5b/controllers/base.go#L54-L60

Then we can easily change the underlying implementation of session data structure.

Comment (Kininaru) [reply controllers/base.go]: @hsluoyz Finished. Now, for each session, there is a struct called `SessionData` stored in session like this: 
```go
type SessionData struct {
	ExpireTime int64
}
```
We can add more info to the struct in the future.

**Review by hsluoyz — APPROVED** [review]

**Pull request merged**